contracts/Identity.sol

```diff
@@ -0,0 +1,396 @@
+     * MUST emit DSNPRemoveDelegate
+     */
+    function delegateRemoveByEIP712Sig(
+        uint8 v,
+        bytes32 r,
+        bytes32 s,
+        DelegateRemove calldata change
+    ) external override {
+        // Checks
+        address signer = delegateRemoveSigner(v, r, s, change);
+
+        // Self removal checks
+        if (!_checkAuthorization(signer, Permission.DELEGATE_REMOVE, block.number)) {
+            require(change.endBlock <= block.number, "Cannot self-remove in the future");
+            require(
+                signer == change.delegateAddr,
+                "Signer does not have the DELEGATE_REMOVE permission"
+            );
+        }
+        require(
+            _delegationData().delegations[change.delegateAddr].nonce == change.nonce,
+            "Nonces do not match"
+        );
+        require(change.nonce > 0, "Never authorized");
+        require(change.endBlock > 0x0, "endBlock 0x0 reserved for endless permissions");
+
+        // Effects
+        _setDelegateEnd(change.delegateAddr, change.endBlock);
+    }
+
+    /**
+     * @dev Get a delegate's nonce
+     * @param addr The delegate's address to get the nonce for
+     *
+     * @return nonce value for delegate
+     */
+    function getNonceForDelegate(address addr) external view override returns (uint32) {
+        return _delegationData().delegations[addr].nonce;
+    }
+
+    /**
+     * @notice Query if a contract implements an interface
+     * @param interfaceID The interface identifier, as specified in ERC-165
+     * @dev Interface identification is specified in ERC-165. This function
+     *  uses less than 30,000 gas.
+     * @return `true` if the contract implements `interfaceID` and
+     *  `interfaceID` is not 0xffffffff, `false` otherwise
+     */
+    function supportsInterface(bytes4 interfaceID) external pure override returns (bool) {
+        return
+            interfaceID == type(ERC165).interfaceId || interfaceID == type(IDelegation).interfaceId;
+    }
+
+    /**
+     * @dev Assigns a delegate role
+     * @param addr The address to assign the given role to
+     * @param role The role to assign
+     */
+    function _setDelegateRole(address addr, Role role) internal {
+        AddressDelegation storage delegation = _delegationData().delegations[addr];
+        delegation.role = role;
+        delegation.endBlock = 0x0;
+        delegation.nonce++;
+        emit DSNPAddDelegate(addr, role);
+    }
+
+    /**
+     * @dev Removes a delegate role at a given point
+     * @param addr The address to revoke the given role to
+     * @param endBlock The exclusive block to end permissions on (0x1 for always)
+     */
+    function _setDelegateEnd(address addr, uint64 endBlock) internal {
+        AddressDelegation storage delegation = _delegationData().delegations[addr];
+        delegation.endBlock = endBlock;
+        delegation.nonce++;
+        emit DSNPRemoveDelegate(addr, endBlock);
```

Comment:
This seems expensive for something that can be viewed in state. Are there existing use cases for an event?

Reply:
It can be viewed in state, but that state can be overridden. That said the transaction has the details of the changes.

Hmm... It is really helpful to have these events for anyone trying to cache the state of delegations however.

I suppose there is no real reason we need DSNPRemoveDelegate or DSNPAddDelegate. Let's discuss at standup.

Reply:
Added an issue over on the spec to discuss this post merge: https://github.com/LibertyDSNP/spec/issues/65